File: scripts/calculate_north_offsets.py
#!/usr/bin/env python3
"""
calculate_north_offsets.py - Calculate northOffset values from quaternion orientations
"""

import json
import math
import numpy as np
import argparse
import os
from pathlib import Path

def quaternion_to_yaw(w, x, y, z):
    """
    Convert quaternion to yaw angle (rotation around Z-axis)
    Returns yaw in degrees
    """
    # Normalize quaternion
    norm = math.sqrt(w*w + x*x + y*y + z*z)
    if norm > 0:
        w, x, y, z = w/norm, x/norm, y/norm, z/norm
    
    # Calculate yaw from quaternion
    # yaw = atan2(2*(w*z + x*y), 1 - 2*(y*y + z*z))
    yaw = math.atan2(2*(w*z + x*y), 1 - 2*(y*y + z*z))
    
    # Convert to degrees
    return math.degrees(yaw)

def calculate_north_offsets(config_file='public/config.json', reference_scene_id=None):
    """
    Calculate northOffset values for all scenes based on their quaternion orientations
    
    Args:
        config_file: Path to the config.json file
        reference_scene_id: ID of the scene to use as reference (0 offset)
                           If None, uses the first scene
    """
    
    # Load configuration
    with open(config_file, 'r') as f:
        config = json.load(f)
    
    scenes = config['scenes']
    
    if not scenes:
        print("No scenes found in configuration")
        return
    
    # Determine reference scene
    if reference_scene_id is None:
        reference_scene = scenes[0]
        reference_scene_id = reference_scene['id']
    else:
        reference_scene = next((s for s in scenes if s['id'] == reference_scene_id), None)
        if not reference_scene:
            print(f"Reference scene {reference_scene_id} not found")
            return
    
    # Calculate reference yaw
    ref_orientation = reference_scene['orientation']
    reference_yaw = quaternion_to_yaw(
        ref_orientation['w'], 
        ref_orientation['x'], 
        ref_orientation['y'], 
        ref_orientation['z']
    )
    
    print(f"Using scene '{reference_scene_id}' as reference (yaw: {reference_yaw:.1f}°)")
    print("\nCalculated northOffset values:")
    print("-" * 50)
    
    # Calculate offsets for all scenes
    for scene in scenes:
        orientation = scene['orientation']
        scene_yaw = quaternion_to_yaw(
            orientation['w'], 
            orientation['x'], 
            orientation['y'], 
            orientation['z']
        )
        
        # Calculate offset relative to reference
        north_offset = scene_yaw - reference_yaw
        
        # Normalize to [-180, 180] range
        while north_offset > 180:
            north_offset -= 360
        while north_offset <= -180:
            north_offset += 360
        
        # Update scene with calculated offset
        scene['northOffset'] = round(north_offset, 1)
        
        print(f"Scene {scene['id']:6s}: yaw={scene_yaw:7.1f}° -> northOffset={north_offset:7.1f}°")
    
    # Save updated configuration
    backup_file = config_file + '.backup'
    print(f"\nCreating backup: {backup_file}")
    
    with open(backup_file, 'w') as f:
        json.dump(config, f, indent=2)
    
    with open(config_file, 'w') as f:
        json.dump(config, f, indent=2)
    
    print(f"Updated {config_file} with calculated northOffset values")
    print(f"\nTotal scenes processed: {len(scenes)}")

def analyze_orientations(config_file='public/config.json'):
    """
    Analyze the orientation distribution of all scenes
    """
    with open(config_file, 'r') as f:
        config = json.load(f)
    
    scenes = config['scenes']
    yaws = []
    
    print("Scene Orientation Analysis:")
    print("-" * 60)
    print(f"{'Scene ID':10s} {'Yaw (°)':>10s} {'Current Offset':>15s}")
    print("-" * 60)
    
    for scene in scenes:
        orientation = scene['orientation']
        yaw = quaternion_to_yaw(
            orientation['w'], 
            orientation['x'], 
            orientation['y'], 
            orientation['z']
        )
        yaws.append(yaw)
        current_offset = scene.get('northOffset', 0)
        
        print(f"{scene['id']:10s} {yaw:10.1f} {current_offset:15.1f}")
    
    print("-" * 60)
    print(f"Yaw range: {min(yaws):.1f}° to {max(yaws):.1f}°")
    print(f"Yaw spread: {max(yaws) - min(yaws):.1f}°")
    print(f"Average yaw: {sum(yaws)/len(yaws):.1f}°")

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Calculate north offsets for panorama scenes')
    parser.add_argument('--project', type=str, help='Project ID for project-specific calculation')
    parser.add_argument('--reference', type=str, help='Reference scene ID to use as 0 offset')
    parser.add_argument('--analyze', action='store_true', help='Analyze current orientations')
    args = parser.parse_args()
    
    # Determine config file path
    if not args.project:
        print('Error: Project ID is required. Use --project <projectId> argument.')
        exit(1)
        
    # Project-specific paths
    config_file = f'public/{args.project}/config.json'
    print(f'Using project-specific config: {config_file}')
    
    if not os.path.exists(config_file):
        print(f'Error: Config file not found at {config_file}')
        exit(1)
    
    if args.analyze:
        analyze_orientations(config_file)
    elif args.reference:
        calculate_north_offsets(config_file, args.reference)
    else:
        calculate_north_offsets(config_file)

File: scripts/node/calculate-north-offsets.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

/**
 * Convert quaternion to yaw angle (rotation around Z-axis)
 * Returns yaw in degrees
 */
function quaternionToYaw(w, x, y, z) {
    // Normalize quaternion
    const norm = Math.sqrt(w*w + x*x + y*y + z*z);
    if (norm > 0) {
        w = w/norm;
        x = x/norm;
        y = y/norm;
        z = z/norm;
    }
    
    // Calculate yaw from quaternion
    // yaw = atan2(2*(w*z + x*y), 1 - 2*(y*y + z*z))
    const yaw = Math.atan2(2*(w*z + x*y), 1 - 2*(y*y + z*z));
    
    // Convert to degrees
    return yaw * 180 / Math.PI;
}

/**
 * Calculate northOffset values for all scenes based on their quaternion orientations
 */
function calculateNorthOffsets(configFile, referenceSceneId = null) {
    // Load configuration
    const config = JSON.parse(fs.readFileSync(configFile, 'utf8'));
    const scenes = config.scenes;
    
    if (!scenes || scenes.length === 0) {
        console.log('No scenes found in configuration');
        return;
    }
    
    // Determine reference scene
    let referenceScene;
    if (referenceSceneId === null) {
        referenceScene = scenes[0];
        referenceSceneId = referenceScene.id;
    } else {
        referenceScene = scenes.find(s => s.id === referenceSceneId);
        if (!referenceScene) {
            console.log(`Reference scene ${referenceSceneId} not found`);
            return;
        }
    }
    
    // Calculate reference yaw
    const refOrientation = referenceScene.orientation;
    const referenceYaw = quaternionToYaw(
        refOrientation.w,
        refOrientation.x,
        refOrientation.y,
        refOrientation.z
    );
    
    console.log(`Using scene '${referenceSceneId}' as reference (yaw: ${referenceYaw.toFixed(1)}°)`);
    console.log('\nCalculated northOffset values:');
    console.log('-'.repeat(50));
    
    // Calculate offsets for all scenes
    for (const scene of scenes) {
        const orientation = scene.orientation;
        const sceneYaw = quaternionToYaw(
            orientation.w,
            orientation.x,
            orientation.y,
            orientation.z
        );
        
        // Calculate offset relative to reference
        let northOffset = sceneYaw - referenceYaw;
        
        // Normalize to [-180, 180] range
        while (northOffset > 180) {
            northOffset -= 360;
        }
        while (northOffset <= -180) {
            northOffset += 360;
        }
        
        // Update scene with calculated offset
        scene.northOffset = Math.round(northOffset * 10) / 10;
        
        console.log(`Scene ${scene.id.padEnd(6)}: yaw=${sceneYaw.toFixed(1).padStart(7)}° -> northOffset=${northOffset.toFixed(1).padStart(7)}°`);
    }
    
    // Save updated configuration
    const backupFile = configFile + '.backup';
    console.log(`\nCreating backup: ${backupFile}`);
    
    fs.writeFileSync(backupFile, JSON.stringify(config, null, 2));
    fs.writeFileSync(configFile, JSON.stringify(config, null, 2));
    
    console.log(`Updated ${configFile} with calculated northOffset values`);
    console.log(`\nTotal scenes processed: ${scenes.length}`);
}

/**
 * Analyze the orientation distribution of all scenes
 */
function analyzeOrientations(configFile) {
    const config = JSON.parse(fs.readFileSync(configFile, 'utf8'));
    const scenes = config.scenes;
    const yaws = [];
    
    console.log('Scene Orientation Analysis:');
    console.log('-'.repeat(60));
    console.log(`${'Scene ID'.padEnd(10)} ${'Yaw (°)'.padStart(10)} ${'Current Offset'.padStart(15)}`);
    console.log('-'.repeat(60));
    
    for (const scene of scenes) {
        const orientation = scene.orientation;
        const yaw = quaternionToYaw(
            orientation.w,
            orientation.x,
            orientation.y,
            orientation.z
        );
        yaws.push(yaw);
        const currentOffset = scene.northOffset || 0;
        
        console.log(`${scene.id.padEnd(10)} ${yaw.toFixed(1).padStart(10)} ${currentOffset.toFixed(1).padStart(15)}`);
    }
    
    console.log('-'.repeat(60));
    console.log(`Yaw range: ${Math.min(...yaws).toFixed(1)}° to ${Math.max(...yaws).toFixed(1)}°`);
    console.log(`Yaw spread: ${(Math.max(...yaws) - Math.min(...yaws)).toFixed(1)}°`);
    console.log(`Average yaw: ${(yaws.reduce((a, b) => a + b, 0) / yaws.length).toFixed(1)}°`);
}

// Command line argument parsing
function parseArgs() {
    const args = process.argv.slice(2);
    const options = {
        project: null,
        reference: null,
        analyze: false
    };
    
    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--project':
                options.project = args[++i];
                break;
            case '--reference':
                options.reference = args[++i];
                break;
            case '--analyze':
                options.analyze = true;
                break;
            case '--help':
                console.log('Usage: node calculate-north-offsets.js --project <projectId> [--reference <sceneId>] [--analyze]');
                console.log('Options:');
                console.log('  --project <projectId>   Project ID for project-specific calculation (required)');
                console.log('  --reference <sceneId>   Reference scene ID to use as 0 offset');
                console.log('  --analyze              Analyze current orientations');
                process.exit(0);
                break;
        }
    }
    
    return options;
}

if (require.main === module) {
    const options = parseArgs();
    
    if (!options.project) {
        console.log('Error: Project ID is required. Use --project <projectId> argument.');
        process.exit(1);
    }
    
    // Project-specific paths
    const configFile = `public/${options.project}/config.json`;
    
    if (!fs.existsSync(configFile)) {
        console.log(`Error: Config file not found at ${configFile}`);
        process.exit(1);
    }
    
    if (options.analyze) {
        analyzeOrientations(configFile);
    } else if (options.reference) {
        calculateNorthOffsets(configFile, options.reference);
    } else {
        calculateNorthOffsets(configFile);
    }
}

module.exports = {
    quaternionToYaw,
    calculateNorthOffsets,
    analyzeOrientations
};